standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/Msck.java

```diff
@@ -229,102 +239,168 @@ public int repair(MsckInfo msckInfo) {
+      success = false;
+    } finally {
+      if (resultOut != null) {
+        try {
+          resultOut.close();
+        } catch (IOException e) {
+          LOG.warn("Failed to close output file: ", e);
+          success = false;
+        }
+      }
     }
-    if (loggedInUser == null) {
-      username = System.getProperty("user.name");
-    } else {
-      username = loggedInUser.getShortUserName();
+    return success;
+  }
+
+  /**
+   * When we add new partitions to a transactional table, we have check the writeIds.
+   * For every newly added partitions, we read the maximum writeId form the directory structure
+   * and compare it to the maximum allocated writeId in the metastore.
+   * If the metastore has never allocated any were are good, the use case would be initialize a table with
+   * existing data. The HMS will be initialized with the maximum writeId. The system will handle every delta directory
+   * as committed ones.
+   * If the writeId is higher in the metastore we can still accept the data, the use case would be after some dataloss
+   * some older data backup was used. The system would able to read the old data.
+   * If however the writeId in the new partition is greater than the maximum allocated in the HMS
+   * we must raise an error. The writedId in the HMS should be increased to match the writeIds in the data files,
+   * but it would most likely cause a lot of problem since the transactional data would become inconsistent
+   * between the HMS and the filesystem.
+   * Further more we need to check for the visibilityTransactionIds written by the compaction.
+   * If we have a higher visibilityId in the directory structure than the current transactionid we need to set
+   * the transactionId sequence higher in the HMS so the next reads may read the content of the
+   * compacted base/delta folders.
+   * @param partsNotInMs partitions only in the FileSystem
+   * @param dbName database name
+   * @param tableName table name
+   * @param txnId actual transactionId
+   */
+  private void validateAndAddMaxTxnIdAndWriteId(Set<CheckResult.PartitionResult> partsNotInMs, String dbName,
+      String tableName, long txnId) throws TException {
+    long maxWriteIdOnFilesystem =
+        partsNotInMs.stream().map(CheckResult.PartitionResult::getMaxWriteId).max(Long::compareTo).orElse(0L);
+    long maxVisibilityTxnId =
+        partsNotInMs.stream().map(CheckResult.PartitionResult::getMaxTxnId).max(Long::compareTo).orElse(0L);
+    validateAndAddMaxTxnIdAndWriteId(maxWriteIdOnFilesystem, maxVisibilityTxnId, dbName, tableName, txnId);
+  }
+
+  private void validateAndAddMaxTxnIdAndWriteId(long maxWriteIdOnFilesystem, long maxVisibilityTxnId, String dbName,
+      String tableName, long txnId) throws TException {
+    long maxAllocatedWriteId = getMsc().getMaxAllocatedWriteId(dbName, tableName);
+    if (maxAllocatedWriteId > 0 && maxWriteIdOnFilesystem > maxAllocatedWriteId) {
```

Comment:
what if we would like to ship FS deltas to the backup HMS due to some replication issue? could you please elaborate what issues do you see in increasing writeId in HMS to match FS?

Reply:
My problem here is, that I don't see how we will end up in a consistent state in that case, when the filesystem will have deltas with writeId-s that are not in the hms. We would start using higher writeIds and the next transactions will start to read those deltas, but how will the compaction work, if there are no transaction information for those writes? who will clean those folders? I think it would become a mess.